daemon/api_systems_test.go

```diff
@@ -160,3 +172,97 @@ func (s *apiSuite) TestGetSystemsNone(c *check.C) {
 
 	c.Assert(sys, check.DeepEquals, &systemsResponse{})
 }
+
+func (s *apiSuite) TestSystemActionRequestInvalid(c *check.C) {
+	type table struct{ label, body, error string }
+	tests := []table{
+		{
+			label: "foobar",
+			body:  `"bogus"`,
+			error: "cannot decode request body into system action:.*",
+		}, {
+			label: "",
+			body:  `{"action":"do","mode":"install"}`,
+			error: "system action requires the system label to be provided",
+		}, {
+			label: "foobar",
+			body:  `{"action":"do"}`,
+			error: "system action requires the mode to be provided",
+		}, {
+			label: "foobar",
+			body:  `{"action":"nope","mode":"install"}`,
+			error: `unsupported action "nope"`,
+		},
+	}
+	for _, tc := range tests {
+		s.vars = map[string]string{"label": tc.label}
+		c.Logf("tc: %#v", tc)
+		req, err := http.NewRequest("POST", "/v2/systems/"+tc.label, strings.NewReader(tc.body))
+		c.Assert(err, check.IsNil)
+		rsp := postSystemsAction(systemsActionCmd, req, nil).(*resp)
+		c.Assert(rsp.Type, check.Equals, ResponseTypeError)
+		c.Check(rsp.Status, check.Equals, 400)
+		c.Check(rsp.ErrorResult().Message, check.Matches, tc.error)
+	}
+}
+
+func (s *apiSuite) TestSystemActionRequestNoSystem(c *check.C) {
+	d := s.daemonWithOverlordMock(c)
+	hookMgr, err := hookstate.Manager(d.overlord.State(), d.overlord.TaskRunner())
+	c.Assert(err, check.IsNil)
+	mgr, err := devicestate.Manager(d.overlord.State(), hookMgr, d.overlord.TaskRunner(), nil)
+	c.Assert(err, check.IsNil)
+	d.overlord.AddManager(mgr)
+
+	s.vars = map[string]string{"label": "20191119"}
+	body := `{"action":"do","mode":"install"}`
+	req, err := http.NewRequest("POST", "/v2/systems/20191119", strings.NewReader(body))
+	c.Assert(err, check.IsNil)
+	rsp := postSystemsAction(systemsActionCmd, req, nil).(*resp)
+
+	c.Assert(rsp.Type, check.Equals, ResponseTypeError)
+	c.Check(rsp.Status, check.Equals, 404)
```

Comment:
Could this test be moved into TestSystemActionRequestInvalid if that would grow a httpStatus field in `type table struct{ label, body, error string }` ? Or is there more to it? (it's fine if not, just curious as the test look very similar to the one that is already a table driven test).

Reply:
Rolled this into a single test with other simple error scenarios.